[SROA] Propagate no-signed-zeros(nsz) fast-math flag on the phi node using function attribute

yashssh opened the pull request:
Its expected that the sequence `return X > 0.0 ? X : -X`, compiled with -Ofast, produces fabs intrinsic. However, at this point, LLVM is unable to do so. 

The above sequence goes through the following transformation during the pass pipeline:
1) SROA pass generates the phi node. Here, it does not infer the fast-math flags on the phi node unlike clang frontend typically does.
2) Phi node eventually gets translated into select instruction. 
Because of missing no-signed-zeros(nsz) fast-math flag on the select instruction, InstCombine pass fails to fold the sequence into fabs intrinsic.

This patch, as a part of SROA, tries to propagate nsz fast-math flag on the phi node using function attribute enabling this folding.

Closes #51601

Co-authored-by: Sushant Gokhale <sgokhale@nvidia.com>

yashssh commented:
The original inspiration for the missed optimization can be seen here https://godbolt.org/z/z3Es5oxqv. 
This seemed to be the most straightforward way to achieve this optimization and a detailed discussion about the issue and why this fold was skipped by 'instcombine' is documented here #51601

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -2742,7 +2742,11 @@ static Instruction *foldSelectWithFCmpToFabs(SelectInst &SI,
+      return nullptr;
+
+    bool functionHasNoSignedZeroes = SI.getParent()->getParent()->hasFnAttribute("no-signed-zeros-fp-math");
+    if(!functionHasNoSignedZeroes && !SI.hasNoSignedZeros())
```
nit: 'functionHasNoSign...' -> 'FunctionHasNoSign...'

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -2742,7 +2742,11 @@ static Instruction *foldSelectWithFCmpToFabs(SelectInst &SI,
+    if (!SI.hasNoNaNs())
+      return nullptr;
+
+    bool functionHasNoSignedZeroes = SI.getParent()->getParent()->hasFnAttribute("no-signed-zeros-fp-math");
```
I think `isFast()` implies `noSignedZeros`, can you just use that?

andykaylor replied:
It's better to look for the individual semantic modes. While `isFast()` implies `noSignedZeroes`, it is possible to enable `noSignedZeroes` without enabling other fast-math flags, so `isFast()` might return false even when `noSignedZeros` is enabled.

andykaylor commented on llvm/test/Transforms/InstCombine/fabs.ll:
```diff
@@ -547,6 +547,20 @@ define double @select_fcmp_nnan_nsz_ult_zero_unary_fneg(double %x) {
 }
 
+
+define float @absfloat32f_olt_fast_no_signed_zeroes(float %x) "no-signed-zeros-fp-math" {
```
If there aren't already test cases showing that this won't happen without the function attribute, can you add them?

yashssh replied:
Added tests with no attributes.

nikic reviewed (commented):
I'm wondering whether we should be propagating the function-level attributes to instruction flags *somewhere* so that this is handled generically, rather than in individual folds?

arsenm commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -2742,7 +2742,12 @@ static Instruction *foldSelectWithFCmpToFabs(SelectInst &SI,
+    if (!SI.hasNoNaNs())
+      return nullptr;
+
+    bool FunctionHasNoSignedZeroes =
+        SI.getParent()->getParent()->hasFnAttribute("no-signed-zeros-fp-math");
```
We shouldn't spread additional uses of the attribute. This should only use the flags

yashssh replied:
I'm unable to get it to work with floating point flags. `instcombine` requires `select` to have `nsz` flag to be able to fold a combination of `fcmp` `fneg` and `select` into a `fabs` intrinsic. But with current semantics, we can't safely propagate `nsz` to `select`. I might not be able to fully explain the problem in single comment but everything is well documented at #51601 particularly [this comment](https://github.com/llvm/llvm-project/issues/51601#issuecomment-981047527).

andykaylor replied:
In principle, I agree with @arsenm about not introducing additional uses of the attribute. I said in #51601 that I couldn't see a way around it, but @nikic's suggestion about setting the flag somewhere based on the attribute might be a way around this. I can't see a reason that it wouldn't be legal for SROA to set the nsz flag when it eliminates the stores and loads in this example: https://godbolt.org/z/fahb8oM9o

arsenm replied:
Unfortunately, this is also the wrong way to test for the attribute. For some reason, these are specified as string booleans. You should use getValueAsBool.

sushgokh replied:
I tried 
```
bool FnHasNszAttr = SI.getFunction()->getFnAttribute("no-signed-zeros-fp-math").getValueAsBool();
```
But this is returning false while below code is returning true
```
bool FnHasNszAttr = SI.getFunction()->hasFnAttribute("no-signed-zeros-fp-math")
```
Is there anything I am misinterpreting?

arsenm commented on llvm/test/Transforms/InstCombine/fabs.ll:
```diff
@@ -839,6 +868,35 @@ define <2 x float> @select_fcmp_nnan_nsz_ugt_zero_unary_fneg(<2 x float> %x) {
+; CHECK-NEXT:    ret float [[RETVAL_0]]
+;
+entry:
+  %cmp = fcmp fast ogt float %x, 0.000000e+00
+  %fneg = fneg fast float %x
```
Reduce set of flags in tests

dtcxzyw commented:
> I'm wondering whether we should be propagating the function-level attributes to instruction flags _somewhere_ so that this is handled generically, rather than in individual folds?

TBH I think fast-math flags should only be function-level attributes :(

yashssh commented:
> > I'm wondering whether we should be propagating the function-level attributes to instruction flags _somewhere_ so that this is handled generically, rather than in individual folds?
> 
> TBH I think fast-math flags should only be function-level attributes :(

Will that be more efficient? What about function inlining when we end up losing these attributes?

yashssh commented:
> I'm wondering whether we should be propagating the function-level attributes to instruction flags _somewhere_ so that this is handled generically, rather than in individual folds?

That can be done but will be a separate work altogether, my understanding is somewhere early in the pipeline. I was also suggested that it would require allowing `loads` and `stores` to have `fast` flags so that they are safely reflected in `phi`s and `select`s.

dtcxzyw commented:
> What about function inlining when we end up losing these attributes?

Is there any real-world project whose translation units/functions are compiled with different fast math flags?

andykaylor commented:
> > What about function inlining when we end up losing these attributes?
> 
> Is there any real-world project whose translation units/functions are compiled with different fast math flags?

We have pragmas that allow you to turn fast-math on an off locally, at a scope level, so even within a function the fast-math flags can change. I think this is an important capability. One of the top concerns I hear from customers who are using fast-math is that they need the ability to track down optimizations that are throwing their results too far out of range. A first step to tracking this down is to stop using fast-math for individual files or sets of files to bisect the problem down to which file is causing a failure. From there, it gets harder, but pragmas can help.

I'd like to introduce a change to the way we check fast-math flags that would let us insert some kind of debug counter to narrow it down in a way that would be similar to how opt-bisect works but effective at the individual transformation granularity. There are some obstacles to making that happen, but I think that would be an extremely beneficial capability. Once you find the place where an accuracy-related failure is being introduced, my idea is to have some way (probably based on pragmas) to turn fast-math off for that location only in the code without needing the debug counter. That's a bit of a pipe dream at this point, but I want to see us working towards that.

This is one of the reasons I don't like fast-math function attributes, but the way things currently work is that the attribute isn't set if the relevant fast-math flag is turned off anywhere in the function. The front end handles this with pragmas, but it means that the inliner needs to remove the attribute if inlining calls where the callee and caller don't both have it set.

andykaylor commented:
> I'm wondering whether we should be propagating the function-level attributes to instruction flags _somewhere_ so that this is handled generically, rather than in individual folds?

The problem is in deciding where to do that. I don't like the function attributes and have made a few attempts to have them deprecated, but there are cases that couldn't be solved without them. This case, as detailed in #51601, is one such example. The front end can't set the fast-math flags on load instructions, so they never make it to the phi instruction that gets created. Propogating the flags from the attribute to instructions wouldn't help unless it was done after SROA and/or SimplifyCFG, but maybe those passes could add the flag based on the function-level attribute. That might be a better way to fix this.

yashssh commented:
I'm leaving this pull request open for now while simultaneously looking if I can set FMF in SROA, will return back with whatever I find.

andykaylor commented:
> I'm leaving this pull request open for now while simultaneously looking if I can set FMF in SROA, will return back with whatever I find.

This is obviously a contrived example, but here is a case that will not be optimized if you rely on the function attributes but could be optimized if the flags were propagated by SROA.

https://godbolt.org/z/senb3bEbn

yashssh commented:
Thanks for the example @andykaylor!

I tried a change that sets `fast` flag on `phi` nodes created by SROA 
```
diff --git a/llvm/lib/Transforms/Utils/PromoteMemoryToRegister.cpp b/llvm/lib/Transforms/Utils/PromoteMemoryToRegister.cpp
index 88b05aab8db4..1d85c7e0d8a6 100644
--- a/llvm/lib/Transforms/Utils/PromoteMemoryToRegister.cpp
+++ b/llvm/lib/Transforms/Utils/PromoteMemoryToRegister.cpp
@@ -1056,6 +1056,22 @@ bool PromoteMem2Reg::QueuePhiNode(BasicBlock *BB, unsigned AllocaNo,
   // BasicBlock.
   PN = PHINode::Create(Allocas[AllocaNo]->getAllocatedType(), getNumPreds(BB),
                        Allocas[AllocaNo]->getName() + "." + Twine(Version++));
+
+  if(BB->getParent()->hasFnAttribute("unsafe-fp-math")){
+    PN->setFast(true);
+  }
+
   PN->insertBefore(BB->begin());
   ++NumPHIInsert;
   PhiToAllocaMap[PN] = AllocaNo;
```
Is this the direction you were hinting towards?

arsenm commented:
> +  if(BB->getParent()->hasFnAttribute("unsafe-fp-math")){
> +    PN->setFast(true);
> +  }
> +

I don't think unsafe-fp-math actually directly corresponds to setting all fast math flags, but it's so poorly defined I think it's impossible to make a conclusive statement 

>    PN->insertBefore(BB->begin());
>    ++NumPHIInsert;
>    PhiToAllocaMap[PN] = AllocaNo;
> ```
> 
> Is this the direction you were hinting towards?

I don't think so. I think the idea was to introduce the flags on the phi based on the context instructions when eliminating the alloca. The whole reason to move it here was to avoid relying on the fast math attributes

andykaylor commented:
> > * if(BB->getParent()->hasFnAttribute("unsafe-fp-math")){
> > * PN->setFast(true);
> > * }
> 
> I don't think unsafe-fp-math actually directly corresponds to setting all fast math flags, but it's so poorly defined I think it's impossible to make a conclusive statement
> 

Although I don't know if this is documented anywhere, I assume that the `"unsafe-fp-math"` attribute is meant to correspond to the -funsafe-math-optimizations command line option. That's carried over from gcc, where the documentation says it "enables -fno-signed-zeros, -fno-trapping-math, -fassociative-math and -freciprocal-math." So for our purposes, that would mean it implies `nsz`, `reassoc`, and `arcp`.

This actually is the direction I had intended, but I think a different approach might be possible. There are some problems that just can't be solved with the current fast-math flags without the function attributes, but I really would like to eliminate the function attributes, so I'd like to explore other ways around this.

I'm not sure I understand what you mean by "introduce the flags on the phi based on the context instructions when eliminating the alloca." What are the "context instructions"? In the case described in #51601, which this PR is trying to address, the `phi` is constructed by SROA combining the load of value produced by an `fneg` instruction (which has the fast-math flags set) and the load of a value stored from a function argument, which currently has no way of expressing the `nsz` flag apart from this function attribute.

If the `phi` were being constructed from two operations that had both had fast-math flags set, SROA should be able to transfer the flags to the `phi`, but it looks like it currently doesn't.

https://godbolt.org/z/P7o3Y6ac7

If we fixed that and added a way to attach fast-math flags to parameters, that might also be a way to solve this. That seems to be the direction you were going with the `nofpclass` attributes. We talked about this a bit in today LLVM FP working group meeting, and @jcranmer-intel expressed general concerns about setting fast-math flags on `select` and `phi` instructions. I think his concerns about the vagueness of the semantics in that case would also apply to using it on a function argument.

Specifically, Joshua's concern is that select and phi don't actually perform an operation, they are just data moving operations, so fast-math shouldn't affect their results. That's a general concern he has, not specific to the `nsz` flag, I think, but the current definition of `nsz` is definitely insufficient. It says "Allow optimizations to treat the sign of a zero argument or zero result as insignificant." That's a bit ambiguous. I think we need it to say something like "Allow optimizations to treat 0.0 and -0.0 as if they were exactly the same value when used as arguments and return values."

The idea here is that this would be similar to the assumption we make everywhere about signaling and quiet NaNs being the same. This would explicitly allow the optimizer to replace a -0.0 result with 0.0 (or vice versa? -- I think there could be problems with that). And in that case, if we had an equivalent parameter attribute, we wouldn't even need the `nsz` flag on the select operation.

Consider the motivating example for this PR:

```
define dso_local double @floatingAbs(double nsz %0)  {
  %2 = fcmp fast olt double %0, 0.000000e+00
  br i1 %2, label %3, label %5

3:                                                ; preds = %1
  %4 = fneg fast double %0
  br label %6

5:                                                ; preds = %1
  br label %6

6:                                                ; preds = %5, %3
  %.0 = phi double [ %4, %3 ], [ %0, %5 ]
  ret double %.0
}
```

I've added a hypothetical `nsz` parameter attribute here for the sake of discussion. Now suppose `%0` is `-0.0` in this case. The `fcmp` instruction will always return `false` when `%0` is `-0.0`, regardless of the fast-math flags, so the `fneg` instruction won't be used and the `phi` instruction will return `%0` in this case. Since the `nsz` attribute is set on the `%0` parameter, we can treat it as though 0.0 and -0.0 were exactly the same value. This allows the llvm.fabs optimization without the `nsz` flag being set on the `phi` instruction.

Does that seem like a reasonable solution?

arsenm commented:
One annoying bit about handling this as a parameter attribute is we can't attach the same to other arbitrary sources (e.g. load) which should behave exactly the same. We would need another mirror metadata

jcranmer-intel commented:
> If we fixed that and added a way to attach fast-math flags to parameters, that might also be a way to solve this. That seems to be the direction you were going with the `nofpclass` attributes. We talked about this a bit in today LLVM FP working group meeting, and @jcranmer-intel expressed general concerns about setting fast-math flags on `select` and `phi` instructions. I think his concerns about the vagueness of the semantics in that case would also apply to using it on a function argument.
> 
> Specifically, Joshua's concern is that select and phi don't actually perform an operation, they are just data moving operations, so fast-math shouldn't affect their results. That's a general concern he has, not specific to the `nsz` flag, I think, but the current definition of `nsz` is definitely insufficient. It says "Allow optimizations to treat the sign of a zero argument or zero result as insignificant." That's a bit ambiguous. I think we need it to say something like "Allow optimizations to treat 0.0 and -0.0 as if they were exactly the same value when used as arguments and return values."

To elucidate this with more detail:
`nsz` (and to a degree `nofpclass`-like stuff in general) are--to a high-level programming language--essentially exposing new variants of floating-point types that have slightly different representations. For `nofpclass`, it's an assertion that certain classes are trap representations; for `nsz`, it considers there to be a single `0` value that has two equivalent representations that is not expected to be stable (almost like NaN payloads, although we would want something weaker than LLVM's payload guarantees). In other words, if LLVM had a `fast_float` type, we wouldn't need any flags on PHIs or selects to enable this optimization, since the type alone would be sufficient to guarantee it.

In lieu of a new floating-point type family, we're essentially inferring the floating-point type of the values based on the flags applied to operations, with bitcasts between the types being implicit. But without flags on select, we have no way to distinguish between `cmp(fast_float x) { x < 0 ? -x : x }` (where the optimization is legal) and `cmp(float x) { (fast_float)x < 0 ? -(fast_float)x : x }` (where it's not legal).

You might get back there by being able to annotate basically any float value with something indicating the desired `nsz` bit on arguments and loads and casts and intrinsics and etc. that let you exhaustively distinguish between `fast_float` and `float` for all operands (there's still several FP instructions right now that can't have fast math flags, like casts). In such a scenario, I think you can have a reliable rule that a `select` or a `phi` where all arguments are `nsz` is itself `nsz`.

sushgokh commented:
1. Lets look at the unoptimized IR for motivating example('nofabs') and its slightly modified version('fabs') here: https://godbolt.org/z/cMe3ca38G
There are 2 aspects to consider here
   a) missing phi node from frontend: For 'nofabs', why are we not getting the phi node as we are getting it for 'fabs'? Modifying clang to generate phi node for 'nofabs' may solve this particular issue.
   
   b) fast math attributes on select/phi instructions: For the 'fabs', front end is probably deducing the fast math attribute from the fast-math flag and already setting it for phi node. In this sense, it wont be inappropriate to set the fast-math flags on the select instruction in 'nofabs' somehow. 
>Joshua's concern is that select and phi don't actually perform an operation, they are just data moving operations

So, I disagree with @jcranmer-intel **for now** considering his statement that select/phi are just data moving ops and that fast-math attributes be not placed on select/phi nodes.

2. Deducing fast-math attributes on the all the instructions just before inlining:
Why not do this similar to 1(b) above and as @nikic might be suggesting?

3. Fast-math attributes on the function parameters:
Is there  documentation somewhere that says if you have fast-math flag enabled at frontend, the fast-math attributes apply to function paramters as well? Also, having `nzero` on function parameter would disallow passing -0.0 as the function argument as the lang reference says:

>"This attribute applies to parameters and return values with floating-point and vector of floating-point types, as well as arrays of such types. The test mask has the same format as the second argument to the llvm.is.fpclass, and indicates which classes of floating-point values are not permitted for the value. For example a bitmask of 3 indicates the parameter may not be a NaN.
If the value is a floating-point class indicated by the nofpclass test mask, a poison value is passed or returned instead."

@andykaylor @jcranmer-intel what do you opine?

sushgokh commented:
ping

sushgokh commented:
There are some additional issues with having attributes on function args(although it may help for this particular case).

Having attributes on the function arguments does not come into picture for the following scenarios:
1. Global variable
```
double x;
double foo()
{
  if(x < 0.0)
    return -x;
  else
    return x;
}
```
2. local variable
```
double foo()
{
   double x;
   scanf(x);
   if (x < 0.0)
        return -x;
    else
        return x;
}
```

Moreover, there is fundamental problem with the IR generated for the motivating example.
```
define double @foo(double %x)
{
entry:
  %cmp = fcmp fast olt double %x, 0.000000e+00     ----------- (1)
  %fneg = fneg fast double %x                      ----------- (2)
  %retval.0 = select i1 %cmp, double %fneg, double %x
  ret double %retval.0
}
```
While nsz flag have been used to enable certain class of optimisations, its presence above in (1) and (2) is actually disabling fabs() optimisation i.e. the meaning of the program is lost with presence of nsz flag. Without nsz flag, as shown below, the meaning is crystal clear and it can easily be turned into fabs() 
```
define double @foo_without_nsz(double %x)
{
entry:
  %cmp = fcmp  olt double %x, 0.000000e+00     
  %fneg = fneg  double %x                      
  %retval.0 = select i1 %cmp, double %fneg, double %x
  ret double %retval.0
}
```
So, one of the things we can consider is removing nsz flag on the  operations that are inherently sensitive to sign of the operand i.e. fcmp, fneg etc.

jcranmer-intel commented:
>        a) missing phi node from frontend: For 'nofabs', why are we not getting the phi node as we are getting it for 'fabs'? Modifying clang to generate phi node for 'nofabs' may solve this particular issue.

I'm not an expert in codegen, but I'm guessing because the ternary operator returns a value, it needs to insert a `phi` to lower the ternary operator correctly, whereas the non-ternary-case is going through variables, which all starts out as `alloca`s.

>        b) fast math attributes on select/phi instructions: For the 'fabs', front end is probably deducing the fast math attribute from the fast-math flag and already setting it for phi node. In this sense, it wont be inappropriate to set the fast-math flags on the select instruction in 'nofabs' somehow.

The broader point here is that when the frontend is generating the PHI nodes, it knows what flags it can set on the operations. But when it is being done by an optimization pass, that knowledge is lost.

>     2. Deducing fast-math attributes on the all the instructions just before inlining:
>        Why not do this similar to 1(b) above and as @nikic might be suggesting?

I don't like the function-based fast-math attributes... but it does seem to be the only quick path out of here.

>     3. Fast-math attributes on the function parameters:
>        Is there  documentation somewhere that says if you have fast-math flag enabled at frontend, the fast-math attributes apply to function paramters as well? Also, having `nzero` on function parameter would disallow passing -0.0 as the function argument as the lang reference says:

`nsz` is by no means equivalent to `nofpclass(nzero)`: in the former case, `-0.0` is treated as equivalent to `+0.0`; in the latter case, `-0.0` is treated as poison, aka, "do whatever you like".

sushgokh commented:
>I don't like the function-based fast-math attributes... but it does seem to be the only quick path out of here.

@jcranmer-intel  could you now review the changes? I have tried addressing the flags issue with the tests.

nikic commented:
Is there anything preventing us from allowing FMF on loads? Or at least the nnan, ninf and nsz flags? (The others don't make much sense.)

yashssh commented:
>Is there anything preventing us from allowing FMF on loads? Or at least the nnan, ninf and nsz flags? (The others don't make much sense.)

That won't solve the problem here, `instcombine` still won't deduce the fast flags correctly on the select instruction as one of the operand is coming from function arguments(IR on right https://godbolt.org/z/P1Pe4vesj ).

Possible solutions discussed in the thread
1) Introduce fast flags on function parameters (long work)
2) Use function attributes to assume it's safe to do such foldings (easy workaround but not the best as pointed out by reviewers)

sushgokh commented:
>Is there anything preventing us from allowing FMF on loads? Or at least the nnan, ninf and nsz flags? (The others don't make much sense.)

Current patch is trying to deduce flags on the instructions from the fast-math attributes on the function and this approach is pretty similar to how clang might be using the cmd-line flags for deducing the instruction level flags.

If we want to deduce flags on loads, for this specific case, we need
1. Fast-math attributes on the function args which might not always be possible especially if `x` is local/global var
2. Have some data flow analysis to propagate those flags.
Correct me if wrong.

If only inlining the issue with using function level attributes, I think that needs to be handled separately.

sushgokh commented:
ping @jcranmer-intel

jcranmer-intel commented:
(Sorry for the delay, I decided it was more prudent to get my FMF RFC out before making any further comments on this PR).

Overall, I don't think this is the right place to do the checking on function-level attributes. Instead, we should probably be having SROA propagate function-level attributes to appropriate FMF (the actual PHI node creation happens in lib/Transforms/Utils/PromoteMemoryToRegister.cpp, not SROA itself), rather than doing one particular query in one particular pattern in InstCombine.

sushgokh commented:
@jcranmer-intel 
>Overall, I don't think this is the right place to do the checking on function-level attributes. Instead, we should probably be having SROA propagate function-level attributes to appropriate FMF

First, its still unclear to me why it is not appropriate to deduce fast-math flags from the func attributes in InstCombine and why that should be done in SROA.

Second, considering above comment, the line of action is unclear.
Consider the SROA output [here](https://godbolt.org/z/MW1z5ecPe) and as can be seen, one of the operands to phi is the function arg. It becomes imperative to have fast-math flags on the func argument and this does not serve the purpose when phi argument is local/global instead of func arg. Could you please elaborate here?

jcranmer-intel commented:
> Second, considering above comment, the line of action is unclear. Consider the SROA output [here](https://godbolt.org/z/MW1z5ecPe) and as can be seen, one of the operands to phi is the function arg. It becomes imperative to have fast-math flags on the func argument and this does not serve the purpose when phi argument is local/global instead of func arg. Could you please elaborate here?

The function-level attributes essentially assert that the entire body of the function was compiled with the `nsz` flag, and thus it would be legal for SROA to add `nsz` to the phi's it generates. It's not about where the phi arguments come from, it's that the phi lives in a function where all code in that function should have the appropriate flag applied to it. (Consequently, when inlining, it's necessary to disable the flag when you inline a function without this attribute into one that has it, but we already do that).

> First, its still unclear to me why it is not appropriate to deduce fast-math flags from the func attributes in InstCombine and why that should be done in SROA.

Reliance on the function-level attributes should probably be minimized as a general concern. As a more practical matter, virtually none of the existing checks for FMF think to check the function-level attributes, and it would be more work to fix all of the existing FMF checks rather than having some pass propagate function-level attributes to the individual instructions that are checked.

In this particular case, we already know that the phis lacking the necessary FMF are coming from SROA, and so it's reasonable to make SROA responsible for propagating function-level attributes to the phis it generates.

nikic commented:
(FWIW, I agree with @jcranmer-intel's last comment. I think the important part is that we don't have any individual InstCombine transform checking function-level flags, and instead propagate the function-level attribute to instruction flags *somewhere* else. And it looks like SROA is a good place to do this as far as this issue is concerned. If we run into more cases where the phi is not created by SROA, we may want an explicit pass to do this.)

jcranmer-intel commented on llvm/lib/Transforms/Utils/PromoteMemoryToRegister.cpp:
```diff
@@ -1112,6 +1113,13 @@ void PromoteMem2Reg::RenamePass(BasicBlock *BB, BasicBlock *Pred,
 
+        if (APN->isComplete() &&
+            APN->getFunction()->hasFnAttribute("no-signed-zeros-fp-math") &&
+            any_of(APN->incoming_values(),
+                   [](Value *V) { return isa<FPMathOperator>(V); })) {
```
`isa<FPMathOperator>(APN)` would work better instead.

jcranmer-intel commented on llvm/test/Transforms/InstCombine/fabs-with-sroa.ll:
```diff
@@ -0,0 +1,140 @@
+; RUN: opt -mtriple=aarch64-unknown-linux-gnu < %s -passes='sroa' -S > %t.ll
+; RUN: cat %t.ll | FileCheck %s --check-prefix=SROA
+; RUN: opt -mtriple=aarch64-unknown-linux-gnu < %t.ll \
+; RUN:     -passes='simplifycfg,instcombine' -S \
```
Testing multiple passes at once is generally discouraged. The test that I'd expect to see here is just that SROA propagates no-signed-zeros-fp-math to the generated PHI, given that we should have tests for the instcombine select/phi working elsewhere.

sushgokh commented:
ping @jcranmer-intel

nikic commented:
The PR description (and title for that matter) could use an update :)

nikic commented on llvm/test/Transforms/SROA/propagate-fast-math-flags-on-phi.ll:
```diff
@@ -0,0 +1,57 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
+; RUN: opt -mtriple=aarch64-unknown-linux-gnu < %s -passes='sroa' -S | FileCheck %s
```
Remove the triple.

nikic commented on llvm/test/Transforms/SROA/propagate-fast-math-flags-on-phi.ll:
It might be worthwhile to also add a PhaseOrdering test that shows that we have end-to-end formation of fabs, as this now involves the interaction of multiple passes.

sushgokh replied:
>Testing multiple passes at once is generally discouraged. The test that I'd expect to see here is just that SROA propagates no-signed-zeros-fp-math to the generated PHI, given that we should have tests for the instcombine select/phi working elsewhere.

I had a test that showed transformation after the passes but this is what @jcranmer-intel suggested. So, I just have a test with SROA transform.

nikic replied:
Adding this kind of test is fine if it tests the whole pipeline (e.g. `-O2`) and is in the PhaseOrdering directory. But if @jcranmer-intel thinks we don't need this, then it's fine.

arsenm commented on llvm/lib/Transforms/Utils/PromoteMemoryToRegister.cpp:
```diff
@@ -1112,6 +1113,12 @@ void PromoteMem2Reg::RenamePass(BasicBlock *BB, BasicBlock *Pred,
           APN->addIncoming(IncomingVals[AllocaNo], Pred);
 
+        if (APN->isComplete() &&
+            APN->getFunction()->hasFnAttribute("no-signed-zeros-fp-math") &&
```
Unfortunately these attributes have explicit boolean parameters. You should use .getValueAsBool() on this. You should add a negative test with an explicit ="false" value

arsenm commented on llvm/test/Transforms/SROA/propagate-fast-math-flags-on-phi.ll:
```diff
@@ -0,0 +1,57 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
+; RUN: opt < %s -passes='sroa' -S | FileCheck %s
+define double @phi_with_nsz(double %x) "no-signed-zeros-fp-math" {
```
the correct form is "no-signed-zeros-fp-math"="true". The verifier should really check this (or we should start auto upgrading to drop the pointless bool value)

sushgokh committed:
[SROA] Propagate no-signed-zeros(nsz) fast-math flag on the phi node using function attribute

Its expected that the sequence return X > 0.0 ? X : -X, compiled with -Ofast, produces fabs intrinsic. However, at this point, LLVM is unable to do so.

The above sequence goes through the following transformation during the pass pipeline:

SROA pass generates the phi node. Here, it does not infer the fast-math flags on the phi node unlike clang frontend typically does.
Phi node eventually gets translated into select instruction.
Because of missing no-signed-zeros(nsz) fast-math flag on the select instruction, InstCombine pass fails to fold the sequence into fabs intrinsic.
This patch, as a part of SROA, tries to propagate nsz fast-math flag on the phi node using function attribute enabling this folding.

Co-authored-by: Sushant Gokhale <sgokhale@nvidia.com>

arsenm commented on llvm/lib/Transforms/Utils/PromoteMemoryToRegister.cpp:
```diff
@@ -1113,6 +1113,14 @@ void PromoteMem2Reg::RenamePass(BasicBlock *BB, BasicBlock *Pred,
+            APN->getFunction()
+                ->getFnAttribute("no-signed-zeros-fp-math")
+                .getValueAsBool() &&
+            isa<FPMathOperator>(APN)) {
+          APN->setHasNoSignedZeros(true);
```
I'd probably reorder these checks; FPOp = dyn_cast<FPMathOperator>(); if (FPop && fnAttrIsNSZ) { FPOp->setHasNoSignedZeros() } 

This also needs a comment explaining why this needs to be done

sushgokh replied:
`FPOp->setHasNoSignedZeros()` was not allowed since setHasNoSignedZeros() is a private member of FPMathOperator

arsenm replied:
It doesn't really make sense that it would be public on Instruction but not on FPMathOperator

arsenm commented on llvm/lib/Transforms/Utils/PromoteMemoryToRegister.cpp:
```diff
@@ -1113,6 +1114,17 @@ void PromoteMem2Reg::RenamePass(BasicBlock *BB, BasicBlock *Pred,
+        // on the phi node generated at this stage, fabs folding does not
+        // happen. So, we try to infer nsz flag from the function attributes to
+        // enable this fabs folding.
+        bool FnHasNSZAttr = APN->getFunction()
+                                ->getFnAttribute("no-signed-zeros-fp-math")
+                                .getValueAsBool();
```
Do this in the expression so it's only lazily checked

sushgokh replied:
you mean check this way, right? 
`if(... && <check nsz flag> && ...)`

I did the above way previously

arsenm commented on llvm/test/Transforms/SROA/propagate-fast-math-flags-on-phi.ll:
```diff
@@ -0,0 +1,55 @@
+return:                          ; preds = %entry,%if.then
+  %retval = load double, ptr %x.addr
+  ret double %retval
+}
```
Also test vector case

sushgokh replied:
why do we need a vector case here? Are you expecting that user starts the pipeline from the vectorization itself and this scalar folding does not happen?

arsenm replied:
Because fp vector typed operations are also FPMathOperator. The pipeline has nothing to do with it, the same pattern should also be applied for any valid IR

sushgokh replied:
I understand but all such vector cases would go through this transformation even before vectorization, right?

arsenm replied:
Vector types can exist in the original source program

arsenm commented on llvm/test/Transforms/SROA/propagate-fast-math-flags-on-phi.ll:
```diff
@@ -0,0 +1,103 @@
+  ret double %retval
+}
+
+define <2 x double> @vector_phi_without_nsz(<2 x double> %x, i1 %cmp, <2 x double> %a, <2 x double> %b) "no-signed-zeros-fp-math"="false" {
```
Don't really need the false vector case

arsenm commented on llvm/lib/Transforms/Utils/PromoteMemoryToRegister.cpp:
```diff
@@ -1113,6 +1114,17 @@ void PromoteMem2Reg::RenamePass(BasicBlock *BB, BasicBlock *Pred,
+        // happen. So, we try to infer nsz flag from the function attributes to
+        // enable this fabs folding.
+        if (APN->isComplete() && isa<FPMathOperator>(APN) &&
+            APN->getFunction()
```
```suggestion
            BB->getParent()
```

arsenm commented:
Should add the Fixes: issue number to the description

sushgokh commented:
ping @arsenm

arsenm commented:
I do think we should have the phase ordering test

sushgokh commented:
> I do think we should have the phase ordering test

Although I didn't have a PhaseOrdering test, I had initially added a test that show the transformation after the concerned passes [here](https://github.com/llvm/llvm-project/pull/83381#discussion_r1603994238) and @jcranmer-intel insisted there that this is not required.

jcranmer-intel commented:
> > I do think we should have the phase ordering test
> 
> Although I didn't have a PhaseOrdering test, I had initially added a test that showed the transformation after the concerned passes [here](https://github.com/llvm/llvm-project/pull/83381#discussion_r1603994238) and @jcranmer-intel insisted there that this is not required.

The right way to do a phase ordering test is not the way you did it. There's a folder test/Transforms/PhaseOrdering directory where the phase-ordering tests go, and in general, you would probably want to check a test like `opt -O1` to make sure that the passes are actually run in the order expected.

arsenm commented on llvm/test/Transforms/PhaseOrdering/fabs.ll:
```diff
@@ -0,0 +1,24 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -passes='default<O1>' -S < %s | FileCheck %s
+
```
Test name could use more elaboration, and a comment

arsenm reviewed (approved):
It's not great but was the consensus on the ticket. The phase orderingtest could use a comment and rename though

nunoplopes commented:
Alive2 complains about this commit:
```llvm
define double @fabs_fcmp_olt_nsz_func_attr(double %#0, double %#1) {
entry:
  %x = alloca i64 8, align 8
  store double %#0, ptr %x, align 8
  %cmp = fcmp nnan nsz olt double %#0, 0.000000
  br i1 %cmp, label %if.then, label %return

if.then:
  %fneg = fneg nnan nsz double %#0
  store double %fneg, ptr %x, align 8
  br label %return

return:
  %ret = load double, ptr %x, align 8
  ret double %ret
}
=>
define double @fabs_fcmp_olt_nsz_func_attr(double %#0, double %#1) {
entry:
  %cmp = fcmp nnan nsz olt double %#0, 0.000000
  br i1 %cmp, label %if.then, label %return

if.then:
  %fneg = fneg nnan nsz double %#0
  br label %return

return:
  %x.0 = phi nsz double [ %fneg, %if.then ], [ %#0, %entry ]
  ret double %x.0
}
Transformation doesn't verify! (unsound)
ERROR: Target's return value is more undefined

Example:
double %#0 = #x0000000000000000 (+0.0)
double %#1 = poison

Source:
ptr %x = pointer(local, block_id=0, offset=0)
i1 %cmp = #x0 (0)
  >> Jump to %return
double %ret = #x0000000000000000 (+0.0)

Target:
i1 %cmp = #x0 (0)
  >> Jump to %return
double %x.0 = #x8000000000000000 (-0.0)
Source value: #x0000000000000000 (+0.0)
Target value: #x8000000000000000 (-0.0)
```

nikic commented:
@nunoplopes As far as I can tell, alive2 does not have support for the `no-signed-zeros-fp-math` attribute, so I think that's expected.

nunoplopes commented:
> @nunoplopes As far as I can tell, alive2 does not have support for the `no-signed-zeros-fp-math` attribute, so I think that's expected.

Ah, thanks, I missed that.
To my defense, it's not documented in LangRef 😅
